### PR TITLE
add case for memory source and allocation

### PR DESCRIPTION
**Pull request by nanli1**
VIRT-297914: Memory source and allocation
Signed-off-by: nanli <nanli@redhat.com>

```
avocado run --vt-type libvirt --test-runner=runner --vt-machine-type q35 memory.backing.source_and_allocation

 (01/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.page_default.file: PASS (53.33 s)
 (02/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.page_default.anonymous: PASS (57.83 s)
 (03/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.page_default.memfd: PASS (58.69 s)
 (04/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.page_default.source_default: PASS (58.68 s)
 (05/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.hugepage.file: PASS (21.93 s)
 (06/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.hugepage.anonymous: PASS (23.30 s)
 (07/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.hugepage.memfd: PASS (23.16 s)
 (08/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.hugepage.source_default: PASS (23.16 s)
 (09/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.page_default.file: PASS (63.34 s)
 (10/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.page_default.anonymous: PASS (52.72 s)
 (11/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.page_default.memfd: PASS (60.55 s)
 (12/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.page_default.source_default: PASS (52.86 s)
 (13/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.hugepage.file: PASS (56.65 s)
 (14/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.hugepage.anonymous: PASS (21.29 s)
 (15/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.hugepage.memfd: PASS (61.14 s)
 (16/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.hugepage.source_default: PASS (50.63 s)
 (17/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.page_default.file: PASS (62.48 s)
 (18/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.page_default.anonymous: PASS (55.55 s)
 (19/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.page_default.memfd: PASS (62.02 s)
 (20/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.page_default.source_default: PASS (53.18 s)
 (21/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.hugepage.file: PASS (22.56 s)
 (22/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.hugepage.anonymous: PASS (19.48 s)
 (23/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.hugepage.memfd: PASS (22.98 s)
 (24/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.hugepage.source_default: PASS (27.06 s)
 (25/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.page_default.file: PASS (63.87 s)
 (26/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.page_default.anonymous: PASS (52.49 s)
 (27/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.page_default.memfd: PASS (63.97 s)
 (28/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.page_default.source_default: PASS (54.99 s)
 (29/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.hugepage.file: PASS (55.86 s)
 (30/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.hugepage.anonymous: PASS (21.86 s)
 (31/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.hugepage.memfd: PASS (61.54 s)
 (32/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.hugepage.source_default: PASS (53.36 s)


```

### CODE REVIEW

**Comment by nanli1** [comment]
modular_redesign

File: libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg [hunk 1]
```diff
@@ -0,0 +1,49 @@
+- memory.backing.source_and_allocation:
+    type = memory_source_and_allocation
+    set_pagesize = 2048
+    set_pagenum = 1024
```
Comment (hs0210) [inline libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg:4]: The default hugepagesize of aarch64 is 512M, please set as follows for aarch64
```
aarch64:
    set_pagesize = 524288
    set_pagenum = 4
````
Test Results based above config:
```
 (01/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.default.file: PASS (35.51 s)
 (02/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.default.anonymous: PASS (38.12 s)
 (03/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.default.memfd: PASS (38.02 s)
 (04/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.default.default: PASS (38.74 s)
 (05/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.hugepage.file: PASS (9.31 s)
 (06/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.hugepage.anonymous: PASS (9.18 s)
 (07/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.hugepage.memfd: PASS (9.32 s)
 (08/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.ondemand.hugepage.default: PASS (9.26 s)
 (09/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.default.file: PASS (39.41 s)
 (10/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.default.anonymous: PASS (39.15 s)
 (11/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.default.memfd: PASS (38.77 s)
 (12/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.default.default: PASS (37.65 s)
 (13/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.hugepage.file: PASS (41.97 s)
 (14/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.hugepage.anonymous: PASS (9.27 s)
 (15/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.hugepage.memfd: PASS (42.88 s)
 (16/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.with_numa.immediate.hugepage.default: PASS (42.79 s)
 (17/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.default.file: PASS (38.99 s)
 (18/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.default.anonymous: PASS (38.50 s)                                                                                                                                                             
 (19/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.default.memfd: PASS (38.69 s)
 (20/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.default.default: PASS (38.44 s)
 (21/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.hugepage.file: PASS (9.33 s)
 (22/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.hugepage.anonymous: PASS (9.20 s)                                                                                                                                                             
 (23/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.hugepage.memfd: PASS (9.25 s)
 (24/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.ondemand.hugepage.default: PASS (9.36 s)
 (25/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.default.file: PASS (39.52 s)
 (26/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.default.anonymous: PASS (38.26 s)                                                                                                                                                            
 (27/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.default.memfd: PASS (38.98 s)
 (28/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.default.default: PASS (38.73 s)
 (29/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.hugepage.file: PASS (41.96 s)
 (30/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.hugepage.anonymous: PASS (9.27 s)                                                                                                                                                            
 (31/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.hugepage.memfd: PASS (41.67 s)
 (32/32) type_specific.io-github-autotest-libvirt.memory.backing.source_and_allocation.memory_allocation.without_numa.immediate.hugepage.default: PASS (41.84 s)
```

Comment (nanli1) [reply libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg]: DONE

File: libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    type = memory_source_and_allocation
+    start_vm = no
+    error_msg = "hugepages are not allowed with anonymous memory source"
+    error_msg_1 = "hugepages are not allowed with memory allocation ondemand"
+    check_allocated_cmd = "grep -A20 '^Size:.*2097152' /proc/`pidof qemu-kvm`/smaps"
+    set_pagesize = 2048
+    set_pagenum = 1024
+    aarch64:
+    set_pagesize = 524288
+    set_pagenum = 4
```
Comment (hs0210) [inline libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg:11]: It may be more readable to indent these two lines.

**Comment by hs0210** [comment]
lgtm

File: libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg (hunk 2 above)
Comment (nanli1) [reply libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg]: Sorry not push success ,Done

**Review by hs0210 — APPROVED** [review]

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+#   Copyright Redhat
+#
+#   SPDX-License-Identifier: GPL-2.0
+
+#   Author: Nan Li <nanli@redhat.com>
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:7]: Please correct your full name, thx!

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        for item in [source_attr, hugepages_attr, alloc_attr]:
+            if item != "":
+                mb_value = mb_value + item + ","
+
+        vm_attrs = eval("{'mb':{%s}}" % mb_value[:-1])
+
+        if numa_attrs:
+            vm_attrs.update(numa_attrs)
+        vm_attrs.update(mem_attrs)
+        test.log.debug("Get current vm attrs is :%s" % vm_attrs)
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:43]:
```suggestion
        test.log.debug("Get current vm attrs is :%s", vm_attrs)
```

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        """
+        expect_xpath = []
+        for xpath in [source_path, hugepages_path, numa_path, mem_xpath,
+                      mem_discard_path]:
+            if xpath != "":
+                expect_xpath.append(eval(xpath))
+        test.log.debug("Get expected xpath: %s", expect_xpath)
+        return expect_xpath
+
+    def get_qemu_cmd_line():
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:60]: Instead of checking qemu-cmd lines, we recommend using some general checking methods, so is it possible to check it in other ways, or is this a mandatory check?

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    def check_allocated_mem(result):
+        """
+        Check the memory allocated from the host.
+
+        :param: result: memory allocated result.
+        """
+        rss_value = int(re.findall(r'Rss:\s+(\w+)', result)[0])
+        private_hugetlb = int(re.findall(r'Private_Hugetlb:\s+(\w+)', result)[0])
+
+        if pagesize == "default" and allocation_mode == "ondemand":
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:99]: Can you pass 'pagesize' and 'allocation_mode' as arguments of this function? It's not easy for me to understand what they do from the definition of this function. What do you think?

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+            if rss_value < 0 or rss_value > mem_value:
+                test.fail("Rss should be ")
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:101]: plz complete the sentence.

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+            if private_hugetlb != 0:
+                test.fail("Private_Hugetlb should be 0 instead of %s", private_hugetlb)
+
+        elif pagesize == "hugepage" and allocation_mode == "ondemand":
+            if rss_value != mem_value:
+                test.fail("Rss should be %s instead of %s", (mem_value, rss_value))
+            if private_hugetlb != 0:
+                test.fail("Private_Hugetlb should be 0 instead of %s",
+                          private_hugetlb)
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:110]:
```suggestion
                test.fail("Private_Hugetlb should be 0 instead of %s" %
                          private_hugetlb)
```

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        elif pagesize == "hugepage" and allocation_mode == "immediate":
+            if rss_value != 0:
+                test.fail("Rss should be 0 instead of %s", rss_value)
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:113]:
```suggestion
                test.fail("Rss should be 0 instead of %s" % rss_value)
```

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+        :param: result: memory allocated result.
+        """
+        rss_value = int(re.findall(r'Rss:\s+(\w+)', result)[0])
+        private_hugetlb = int(re.findall(r'Private_Hugetlb:\s+(\w+)', result)[0])
+
+        if pagesize == "default" and allocation_mode == "ondemand":
+            if rss_value < 0 or rss_value > mem_value:
+                test.fail("Rss should be ")
+            if private_hugetlb != 0:
+                test.fail("Private_Hugetlb should be 0 instead of %s", private_hugetlb)
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:103]:
```suggestion
                test.fail("Private_Hugetlb should be 0 instead of %s" % private_hugetlb)
```

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+        elif pagesize == "hugepage" and allocation_mode == "ondemand":
+            if rss_value != mem_value:
+                test.fail("Rss should be %s instead of %s", (mem_value, rss_value))
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:107]:
```suggestion
                test.fail("Rss should be %s instead of %s" % (mem_value, rss_value))
```

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 8 above)
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:110]:
```suggestion
                test.fail("Private_Hugetlb should be 0 instead of %s" %
                          private_hugetlb)
```

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+            if private_hugetlb != 0:
+                test.fail("Private_Hugetlb should be 0 instead of %s",
+                          private_hugetlb)
+        elif pagesize == "hugepage" and allocation_mode == "immediate":
+            if rss_value != 0:
+                test.fail("Rss should be 0 instead of %s", rss_value)
+            if private_hugetlb != mem_value:
+                test.fail("Private_Hugetlb should be %s instead of %s",
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:115]:
```suggestion
                test.fail("Private_Hugetlb should be %s instead of %s" %
```

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+        Start vm and check config
+        Check the qemu cmd line
+        Check the memory allocated
+        Login the guest and consume the guest memory
+        """
+        test.log.info("TEST_STEP1: Define vm and check result")
+        vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+        vm_attrs = get_vm_attrs()
+        vmxml.setup_attrs(**vm_attrs)
+        test.log.debug("Define vm with %s." % vmxml)
```
Comment (Yingshun) [inline libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:136]:
```suggestion
        test.log.debug("Define vm with %s.",  vmxml)
```

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 5 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py:60]: hi yingshun, Thanks for your patience, But this is to get expected qemu cmd line from different scenario instead of checking qemu cmd line. Do I get what you mean?

**Comment by nanli1** [comment]
Need more time to wait for feature owner checking if  the qemu cmd line method would be replaced by other methods

**Comment by Yingshun** [comment]
> Need more time to wait for feature owner checking if the qemu cmd line method would be replaced by other methods

@liang-cong-red-hat  Could you plz check if we have other ways to check the functionality here instead of checking qemu cmd ASAP?
 
It has been more than 2 months since @nanli1 submitted this pr, and we have a guideline that our prs should be closed within 3 months. Thx!

**Comment by liang-cong-red-hat** [comment]
yeah，since the new way has to do the following 2 steps:
1. qemu cmd line check step has to split to several steps according to different attributes verification.
2. different attributes needs to be run under all scenarios.
I would finish it in this week

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 3 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 4 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 5 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 6 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 7 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 8 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 9 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 11 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 8 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 12 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py (hunk 13 above)
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun ,Done

File: libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg [hunk 14]
```diff
@@ -0,0 +1,57 @@
+- memory.backing.source_and_allocation:
+    type = memory_source_and_allocation
+    start_vm = no
+    error_msg = "hugepages are not allowed with anonymous memory source"
+    error_msg_1 = "hugepages are not allowed with memory allocation ondemand"
```
Comment (Yingshun) [inline libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg:5]: May I ask why you need two variants for these error messages? are they error messages for different scenarios, can you use one? And do you need to set them for all the scenarios?

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+        backend_type = json.loads(result)['return']
+
+        res = virsh.qemu_monitor_command(
+            vm_name, check_thread % (mem_name, 'mem-path'), debug=True).stdout_text
+
+        expected_backend_type, expected_mem_path = '', ''
+
+        if source_set == "file" and pagesize_set == "page_default":
+            expected_backend_type = "memory-backend-file"
+            expected_mem_path = "/var/lib/libvirt/qemu/ram/"
```
[the remark this answers is not in the document]
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks your help yingshun, This helps a lot ,Done

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+                    expected_mem_path, mem_path))
+
+        elif source_set in ["file", "source_default"] and pagesize_set == "hugepage":
+            expected_backend_type = "memory-backend-file"
+            expected_mem_path = "/dev/hugepages/libvirt/qemu/"
+            # Check memory path
+            mem_path = json.loads(res)['return']
+            if expected_mem_path not in mem_path:
+                test.fail("Expected memory path '%s', but got '%s'" % (
+                    expected_mem_path, mem_path))
```
[the remark this answers is not in the document]
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: One is removed due to the next comment you gived :D

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+        """
+        Setup pagesize
+        """
+        hp_cfg.set_kernel_hugepages(set_pagesize, set_pagenum)
+
+    def run_test():
+        """
+        Define vm and check msg
+        Start vm and check config
+        Check the qemu cmd line
```
[the remark this answers is not in the document]
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: DONE

File: libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+            if xpath != "":
+                expect_xpath.append(eval(xpath))
+        test.log.debug("Get expected xpath: %s", expect_xpath)
+        return expect_xpath
+
+    def check_allocated_mem(params, result):
+        """
+        Check the memory allocated from the host.
+
+        :param: result: memory allocated result.
```
[the remark this answers is not in the document]
Comment (nanli1) [reply libvirt/tests/src/memory/memory_backing/memory_source_and_allocation.py]: Thanks yingshun,DONE

File: libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg (hunk 14 above)
Comment (nanli1) [reply libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg]: Thanks yingshun, DONE

Comment (Yingshun) [reply libvirt/tests/cfg/memory/memory_backing/memory_source_and_allocation.cfg]: Do you need to set 'error_msg' for all the scenarios? So all test scenarios in this test are negative tests, right?

**Review by Yingshun — APPROVED** [review]
